Bug fixes in profiling allocator

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45993 Bug fixes in profiling allocator**

Summary:
Some bug exposed via updated test and validation code.
Also enabled this test to be run on CI instead of just mobile only test.

Test Plan:
cpu_profiling_allocator_test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24172599](https://our.internmc.facebook.com/intern/diff/D24172599)

kimishpatel committed:
Bug fixes in profiling allocator

Summary:
Some bug exposed via updated test and validation code.
Also enabled this test to be run on CI instead of just mobile only test.

Test Plan:
cpu_profiling_allocator_test

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

dzhulgakov commented on c10/mobile/CPUProfilingAllocator.cpp:
```diff
@@ -49,6 +31,36 @@ struct MemEvent {
+    auto start_offset = allocation_offsets[alloc_id];
+    auto end_offset = allocation_offsets[alloc_id] + event.size;
+    if (event.type == EventType::Allocate) {
+      if (!allocations.emplace(start_offset, end_offset).second) {
```
if you really want to validate - you should check that the inserted block doesn't overlap with existing blocks here

kimishpatel replied:
@dzhulgakov Thats what this emplace is supposed to do. Did I miss something?

dzhulgakov replied:
Emplace will only complain if the object is equal (which in this case is probably expressed via `!(a<b) && !(b<a)`). I originally assumed that you just sort the items in the set using their `start_offsets`, but it looks like you actually define the partial order above: `    return end_offset <= other.start_offset;	`.

I think it's technically not correct to create `std::set` with such comparison operator. STL documentation requires "strict weak ordering" (http://www.cplusplus.com/reference/set/set/), i.e. https://en.wikipedia.org/wiki/Weak_ordering#Strict_weak_orderings . In particular string weak ordering requires "transitivity of incompatibility":

> For all x, y, z in S, if x is incomparable with y (neither x < y nor y < x hold), and y is incomparable with z, then x is incomparable with z (transitivity of incomparability).

However, you operator doesn't satisfy it, it's a non-strict weak ordering, e.g.:

```
x = (1, 10)
y = (5, 15)
z = (11, 20)
x < y || y < x == False # x and y are incompatible
y < z || z < y == False # y and z are incompatible
x < z # boom!
```

What it means that under certain overlapping circumstances you can get crash/assertion/runtime error being thrown from std::set operation.

I'd also give you another reason to not do it: it's very non-obvious to rely on such fancy comparison operator and it's very likely the reader of this code will overlook it (like I did).

To summarize: please change comparison operator to just look at starts only (so that it's a strict ordering) and then add necessary checks here that the inserted element doesn't overlap with the existing ones. It should be sufficient to just check the iterator just after the inserted one and verify that its start is after the inserted end.

kimishpatel replied:
@dzhulgakov, thanks for pointing this out. I did miss "strict weak ordering" part and thanks for the explanation. That does make sense. Also for readability part, I do agree. It does get a bit cryptic. I should have put comments there at the least.

I will make changes.

Thanks once again. This was informative.

kimishpatel committed:
Update on "Bug fixes in profiling allocator"


Summary:
Some bug exposed via updated test and validation code.
Also enabled this test to be run on CI instead of just mobile only test.

Test Plan:
cpu_profiling_allocator_test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24172599](https://our.internmc.facebook.com/intern/diff/D24172599)

[ghstack-poisoned]

dzhulgakov commented on c10/mobile/CPUProfilingAllocator.cpp:
```diff
@@ -43,12 +51,33 @@ bool validate_allocation_plan(
+          end_offset, std::numeric_limits<uint64_t>::max());
+      auto it = allocations.lower_bound(query_block);
+      if (it != allocations.end()) {
+        while(it != allocations.begin()) {
```
you don't need to iterate actually. You just need to check two blocks - one just before (`--it`) and the one just after (`it`). Because the blocks in the set are note overlapping, it's sufficient

kimishpatel replied:
Yes thats right. I think I forgot the invariant that this set will always only contain non-overlapping intervals. With that invariant you need to check maximum of two other intervals.

kimishpatel replied:
Also just a clarifying comment: candidate_block = (some_start, some_end) while the block I was using to perform query was (end_offset, max) so the search will point you the first block with start_offset >= some_end. However give the other comment regarding invariant, this does not matter anymore.

dzhulgakov commented on c10/mobile/CPUProfilingAllocator.cpp:
```diff
@@ -49,6 +31,65 @@ struct MemEvent {
+      }
+      allocations.emplace(mem_block);
+    } else if (event.type == EventType::Free) {
+      auto it = allocations.find(mem_block);
```
nit: you also need to validate the end of the range here. Because of your comparison operator `find` only finds the element based on the start

kimishpatel replied:
So based on the previous comment regarding the invariant that the set contains only non-overlapping intervals, check based on start offset should suffice. And given allocations should always precede free, if we have found overlapping interval we would have returned from this function beforehand.
Although I might have missed something again.

dzhulgakov replied:
I wasn't referring to overlaps here, I was referring to sanity checking that the end of the interval you have in the set is the same as the end of the interval you're trying to remove.

dzhulgakov commented on c10/mobile/CPUProfilingAllocator.cpp:
```diff
@@ -49,6 +31,65 @@ struct MemEvent {
+      const MemBlock query_block(
+          end_offset, std::numeric_limits<uint64_t>::max());
+      auto it = allocations.lower_bound(query_block);
+      if (it != allocations.end()) {
```
you still need to check the previous block even if it IS the end (as long as it's not the beginning)

kimishpatel replied:
Thanks for that catch.

kimishpatel committed:
Update on "Bug fixes in profiling allocator"


Summary:
Some bug exposed via updated test and validation code.
Also enabled this test to be run on CI instead of just mobile only test.

Test Plan:
cpu_profiling_allocator_test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24172599](https://our.internmc.facebook.com/intern/diff/D24172599)

[ghstack-poisoned]

kimishpatel committed:
Update on "Bug fixes in profiling allocator"


Summary:
Some bug exposed via updated test and validation code.
Also enabled this test to be run on CI instead of just mobile only test.

Test Plan:
cpu_profiling_allocator_test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24172599](https://our.internmc.facebook.com/intern/diff/D24172599)

[ghstack-poisoned]

dzhulgakov commented on c10/mobile/CPUProfilingAllocator.cpp:
```diff
@@ -49,6 +31,63 @@ struct MemEvent {
+    MemBlock mem_block(start_offset, end_offset);
+    if (event.type == EventType::Allocate) {
+      auto it = allocations.lower_bound(mem_block);
+      if (!allocations.empty()) {
```
you don't need this if statement

dzhulgakov commented on c10/mobile/CPUProfilingAllocator.cpp:
```diff
@@ -49,6 +31,63 @@ struct MemEvent {
+          if (overlaps(next_block, mem_block)) {
+            return false;
+          }
+          if (it != allocations.begin()) {
```
just pull it out of the surrounding `if` statement on `end` - your `else` part does exactly the same thing

dzhulgakov reviewed (approved):
Please address comments before landing

kimishpatel committed:
Update on "Bug fixes in profiling allocator"


Summary:
Some bug exposed via updated test and validation code.
Also enabled this test to be run on CI instead of just mobile only test.

Test Plan:
cpu_profiling_allocator_test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24172599](https://our.internmc.facebook.com/intern/diff/D24172599)

[ghstack-poisoned]

kimishpatel committed:
Update on "Bug fixes in profiling allocator"


Summary:
Some bug exposed via updated test and validation code.
Also enabled this test to be run on CI instead of just mobile only test.

Test Plan:
cpu_profiling_allocator_test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24172599](https://our.internmc.facebook.com/intern/diff/D24172599)

[ghstack-poisoned]

kimishpatel commented:
@dzhulgakov Can you also approve the diff: D24172599

kimishpatel committed:
Update on "Bug fixes in profiling allocator"


Summary:
Some bug exposed via updated test and validation code.
Also enabled this test to be run on CI instead of just mobile only test.

Test Plan:
cpu_profiling_allocator_test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24172599](https://our.internmc.facebook.com/intern/diff/D24172599)

[ghstack-poisoned]

kimishpatel committed:
Update on "Bug fixes in profiling allocator"


Summary:
Some bug exposed via updated test and validation code.
Also enabled this test to be run on CI instead of just mobile only test.

Test Plan:
cpu_profiling_allocator_test

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24172599](https://our.internmc.facebook.com/intern/diff/D24172599)

[ghstack-poisoned]
